design/ceph-mon-pv.md

```diff
@@ -0,0 +1,139 @@
+volume mounts. This will continue to be based on `dataDirHostPath`.
+
+## Host networking
+
+Currently Ceph monitor pods may be deployed using host-based networking, or
+placed behind a service using a portable network address. This poses a challenge
+when migrating a monitor (and it's PV) to a new node which has been deployed
+using host-based networking as the monitor's address will no longer valid.
+
+Support for migration of monitors with host-based networking is future work. To
+simplify the initial implementation, host-based networking will not be supported
+for non-local PVs.
+
+# StatefulSet pod management
+
+Monitor pods will be managed by a StatefulSet rather than Deployments which are
+currently use. This change offers (1) simplified naming for DNS and pods (e.g.
+`ceph-mon-a` rather than `ceph-mon-a-3ab423fb`), and (2) StatefulSets are able
+to handle much of the complexity of managing PVs including the mechanics of
+PV/pod association and scheduling.
+
+StatefulSets will be used to manage monitors pods in a 1-1 mapping between a pod
+and StatefulSet. This choice is made to retain maximum flexibility on naming,
+placement, and life cycle management of the pods by Rook.
+
+## Scheduling
+
+Rook currently makes explicit scheduling decisions for monitors. Standard node
+selector is used to control placement for non-local PVs (or dynamically
+provisions local PVs [0]). For statically provisioned local PVs, the Rook
+monitor scheduler will likely need to be made aware of available PVs.
+
+If we want to continue to support the current behavior of Rook behavior
+effectively like a HostPath dynamic provisioner then we could continue to use
+the HostPath mechanism, but we could also allow Rook to create the local PV
+itself. Effectively like this:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: ceph-mon-a-pv
+  labels:
+    vol: ceph-mon-a-pv
+spec:
+  capacity:
+    storage: 1Gi
+  volumeMode: Filesystem
+  accessModes:
+    - ReadWriteOnce
+  storageClassName: local-storage
+  local:
+    path: /opt/mon-a-0-0
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: kubernetes.io/hostname
+          operator: In
+          values:
+            - worker1
+---
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: local-storage
+provisioner: kubernetes.io/no-provisioner
+volumeBindingMode: WaitForFirstConsumer
+```
+
+# Upgrades
+
+Because of the switch from Deployments to Stateful sets, the most
+straightforward method for upgrading the cluster will be to perform a
+controlled replacement of each existing monitor deployment.
+
+There will be other upgrade scenarios that need to be addressed but as long
+as networking and paths between an existing Deployment and the replacement
+StatefulSet are compatible this should be possible, even if it would require
+a file system deep copy of the monitor data path.
```

Comment:
Should we consider a special job which handles the filesystem copy from current-mode-Rook to StatefulSet mons? It might be fairly simple really: make a job that has the same hostPath connection as the mon as well as a connection to the PV, and then `cp -a /host/var/lib/ceph-<id>/mon/ /pv/var/lib/ceph-<id>/mon/`?

Reply:
Yeh, I can imagine some scenarios where this can be further optimized to local file system operations, but generally I think having a single Job handle all the cases, even if some are non-optimal, in order to avoid special casing migration just for a few GBs makes sense.

Reply:
I'd vote for leaving compatibility in for the current mons, and only start the new mons on PVs when mon failover happens. Then it's a simple documentation of using the mon failover feature if people want to start using PVs. It may be more complex to manage a mix of the mons, but it seems simpler than trying to migrate them.

Reply:
Travis might have a good point here. As long as it's not overly complex to support mixed cases, this does seem easier.

Reply:
I'd like to hone in on a specific case. Moving from the existing HostPath approach based on Deployment to StatefulSets. A few upgrade scenarios:

1. No change, continue to use Deployment
2. Move to StatefulSet with HostPath
3. Move to StatefulSet with local PV (path-based)
4. Move to StatefulSet with PV (local _device_-based, OR non-local PV)

In versions (2) and (3) above it seems sufficient to move to StatefulSet because the backing store can remain the same path on the host. Only 4 would require either using the fail over method or a job to perform the migration.

Reply:
Assuming there is a non-disruptive path (2 and 3), sounds great to move the mons to statefulsets.